package/MDAnalysis/lib/include/batched_distances.h

```diff
@@ -0,0 +1,179 @@
+ style API, with contiguity of memory access of paramount importance.
+ Optimal sizing of the batching is yet to be fully explored.
+
+ Functions in this header accept the iterators defined in `iterators.h`
+ (currently _AtomGroupIterator and _ArrayIterator) that provide a homogenous
+ interface for iterating through their respective datatypes
+ (AtomGroups and NumPy ndarrays).
+
+ Templates are used to allow iterators at either position, avoiding writing
+ multiple overloads for functions that accept many iterators.
+
+*/
+
+/* _calc_distance_array_batched uses the following batched algorithm:
+   - figure out how many perfect tiles of size (batchsize, batchsize) we can do
+   - figure out how overhanging our remaining distances are
+   - figure out the maximum number of coordinates at a time for overhang (gcd)
+   - calculate distances in tiles
+   - if overhanging in ref dimension, do overhang NOTE: non contiguous access
+   - if overhanging in conf dimension, do overhang
+
+   NOTE: possible poor performance on overhanging portion when gcd is low (ie a
+   prime number of distances). Should be investigated.
+ */
+
+
+// implement gcd because we don't compile with C++17
+template<typename T>
+T _gcd(T a, T b) {
+   if (b == 0)
+   return a;
+   return _gcd(b, a % b);
+}
+
+template <typename T, typename U>
+void _calc_distance_array_batched(T ref, U conf, double *distances, uint64_t batchsize)
+{
+    // batchsize is in particles, so allocate buffers of 3x batchsize.
+    const uint64_t atom_bufsize = 3 * batchsize;
+    float ref_buffer[atom_bufsize];
+    float conf_buffer[atom_bufsize];
+
+    // slight indirection required here to homogenize interface between
+    // _AtomGroupIterator and _ArrayIterator where passing stack allocated
+    // array as float*& does not decay to a float* as desired.
+    float *ref_buffer_ = ref_buffer;
+    float *conf_buffer_ = conf_buffer;
+
+    uint64_t nref = ref.n_atoms;
+    uint64_t nconf = conf.n_atoms;
+
+    // check if batchsize is larger than the number of coordinates and if so
+    // truncate.
+    uint64_t bsize_ref = std::min(nref, batchsize);
+    uint64_t bsize_conf = std::min(nconf, batchsize);
+
+    // internals
+    double rsq;
+    float dx[3];
+    uint64_t iter_ref;
+    uint64_t iter_conf;
+
+    // what is modulo number of particles?
+    uint64_t ref_overhang = nref % bsize_ref;
+    uint64_t conf_overhang = nconf % bsize_conf;
+
+    // what is the gcd to do iteration of overhang?
+    // using a gcd allows us to use the maximum size of memory buffer that fits
+    // remaining iteration in (gcd_ref, gcd_conf) sized tiles
+    int64_t gcd_conf = _gcd(nconf, bsize_conf);
+    int64_t gcd_ref = _gcd(nref, bsize_ref);
+
+    // note i is used as index of ref and j is used as index of conf.
+
+    for (iter_ref = 0; iter_ref < nref - ref_overhang; iter_ref += bsize_ref)
+    {
+        ref.load_into_external_buffer(ref_buffer_, bsize_ref);
```

Comment:
So to avoid the two loops, one for regular and one for overhang, how about the following:
```c++
for (int i=0; i<n_ref; ++i) {
    if !(i % bsize_ref) {
        ref.load_into_external_buffer(ref_buffer, i, bsize_ref);
        local_i = 0;
    }

    for (int j=0; j<nconf; ++j) {
        if !(j % bsize_conf) {load stuff}


        dx[0] = conf_buffer_[3 * local_j] - ref_buffer_[3 * local_i];
        dx[1] = conf_buffer_[3 * local_j + 1] - ref_buffer_[3 * local_i + 1];
        dx[2] = conf_buffer_[3 * local_j + 2] - ref_buffer_[3 * local_i + 2];
        rsq = (dx[0] * dx[0]) + (dx[1] * dx[1]) + (dx[2] * dx[2]);
        *(distances + nconf * i + j) = sqrt(rsq);
    }
}
```

So
- you loop over `i` and `j` as you normally would, and every `bsize` iterations you refill the buffers.
- there's a new variable `local_i` which tracks the position within a buffer.  This is identical to `i % bsize` (and if bsize_ref is a clever number this is therefore `i & 1023` or similar)
- for the above reason I'd make `bsize` a constant somewhere so the compiler does things (but maybe you know this and are optimizing this later)
- The signature to `load_into_external_buffer` changes to accept a start point for the load, which removes the need for it to track any state (this is instead done inside the function stack)

Reply:
Just for clarity I will paraphrase a private conversation between myself and @richardjgowers  here. With my current implementation this approach doesn't allow us to utilise the so called "squareness" of the computation. i.e.

```c++
for (int i=0; i<n_ref; ++i) {
    if !(i % bsize_ref) {
        ref.load_into_external_buffer(ref_buffer, i, bsize_ref);
        local_i = 0;
    }

    for (int j=0; j<nconf; ++j) { // <- HERE j=0 and j%bsize_conf is true every i
    // so we reload the buffer way too often 
        if !(j % bsize_conf) {load stuff}

```
Which means we are refilling one buffer for every iteration on one coordinate. There is likely a way around this but it didn't jump out to me right away. 

A secondary concern is that with a single loop only and no separation for overhang you need to explicitly check that you are not overrunning the edge of the total number of indices. i.e.

```c++
    if !(i % bsize_ref) {
        ref.load_into_external_buffer(ref_buffer, i, bsize_ref); // <- this can read over the total number of idx 
        local_i = 0;
    }
```
thus requiring an explicit check in the loading function which is slow. 

As a result of this I have reverted to my previous design. I know its a bit ugly with three large nested loops, but it is as fast as `_calc_distance_array()` or faster according to my benchmarking  and gives us the flexibility that we want.